### ISSUE TITLE
Can't pickle local object 'DistributedDataParallel._register_nccl_grad_hook.

### ISSUE THREAD

**Issue by PetrochukM** [pytorch/pytorch#11683]
## Issue description

```
$ python3 -m torch.distributed.launch --nproc_per_node=2 --master_port=1234 abcd.py
Traceback (most recent call last):
  File "abcd.py", line 18, in <module>
    torch.save(model, 'model.pt')
  File "/home/michaelp/.local/lib/python3.6/site-packages/torch/serialization.py", line 209, in save
    return _with_file_like(f, "wb", lambda f: _save(obj, f, pickle_module, pickle_protocol))
  File "/home/michaelp/.local/lib/python3.6/site-packages/torch/serialization.py", line 134, in _with_file_like
    return body(f)
  File "/home/michaelp/.local/lib/python3.6/site-packages/torch/serialization.py", line 209, in <lambda>
    return _with_file_like(f, "wb", lambda f: _save(obj, f, pickle_module, pickle_protocol))
  File "/home/michaelp/.local/lib/python3.6/site-packages/torch/serialization.py", line 282, in _save
    pickler.dump(obj)
AttributeError: Can't pickle local object 'DistributedDataParallel._register_nccl_grad_hook.<locals>.allreduce_hook'
Traceback (most recent call last):
  File "abcd.py", line 18, in <module>
    torch.save(model, 'model.pt')
  File "/home/michaelp/.local/lib/python3.6/site-packages/torch/serialization.py", line 209, in save
    return _with_file_like(f, "wb", lambda f: _save(obj, f, pickle_module, pickle_protocol))
  File "/home/michaelp/.local/lib/python3.6/site-packages/torch/serialization.py", line 134, in _with_file_like
    return body(f)
  File "/home/michaelp/.local/lib/python3.6/site-packages/torch/serialization.py", line 209, in <lambda>
    return _with_file_like(f, "wb", lambda f: _save(obj, f, pickle_module, pickle_protocol))
  File "/home/michaelp/.local/lib/python3.6/site-packages/torch/serialization.py", line 282, in _save
    pickler.dump(obj)
AttributeError: Can't pickle local object 'DistributedDataParallel._register_nccl_grad_hook.<locals>.allreduce_hook'
```

## Code example

```
import argparse
import torch

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument("--local_rank", type=int)
    args = parser.parse_args()

    device = torch.device('cuda', args.local_rank)

    torch.distributed.init_process_group(backend='nccl')
    model = torch.nn.LSTM(10, 10).to(device)
    model = torch.nn.parallel.DistributedDataParallel(
        model, device_ids=[args.local_rank], output_device=args.local_rank, dim=1)

    torch.save(model, 'model.pt')
```

## System Info

PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-16ubuntu3) 7.3.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Tesla P100-PCIE-16GB
GPU 1: Tesla P100-PCIE-16GB
GPU 2: Tesla P100-PCIE-16GB
GPU 3: Tesla P100-PCIE-16GB

Nvidia driver version: 390.30
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.3

**Comment by ezyang** [comment]
ISTR I turned on hook serialization. I guess it's undesirable. Should be an easy fix.

**Comment by ezyang** [comment]
@PetrochukM Could you please try this patch and see if it resolves the problem?

```
diff --git a/torch/multiprocessing/reductions.py b/torch/multiprocessing/reductions.py
index a346d5693..8b3c40222 100644
--- a/torch/multiprocessing/reductions.py
+++ b/torch/multiprocessing/reductions.py
@@ -173,9 +173,9 @@ def reduce_tensor(tensor):
                  handle,
                  storage_size,
                  tensor.requires_grad,
-                 tensor._backward_hooks))
+                 None))
 
-    metadata = (tensor.storage_offset(), tensor.size(), tensor.stride(), tensor.requires_grad, tensor._backward_hooks)
+    metadata = (tensor.storage_offset(), tensor.size(), tensor.stride(), tensor.requires_grad, None)
     return (rebuild_tensor, (type(tensor), storage, metadata))
 
 
```

**Comment by ezyang** [comment]
I confirmed that the patch does NOT solve this error, because the serialization path here is actually NOT exercising the multiprocessing serializer.

**Comment by ezyang** [comment]
Oh, actually, the problem is the new c10d module is not serializable. @teng-li has filed a task about it.

**Comment by ezyang** [comment]
Filed public bug report at #12168

**Comment by ezyang** [comment]
Here's the diff that *actually* fixes the problem, no actually I swear:

```
diff --git a/torch/nn/parameter.py b/torch/nn/parameter.py
index 1035b0705..1169fff0f 100644
--- a/torch/nn/parameter.py
+++ b/torch/nn/parameter.py
@@ -30,5 +30,5 @@ class Parameter(torch.Tensor):
     def __reduce_ex__(self, proto):
         return (
             torch._utils._rebuild_parameter,
-            (self.data, self.requires_grad, self._backward_hooks)
+            (self.data, self.requires_grad, None)
         )
```

but I need to work out if this is the right thing to do.

**Comment by ezyang** [comment]
OK, after some discussion with @apaszke, we've decided that:

1. Under *no* circumstances should we serialize hooks. This is a BC-breaking change, but it's consistent with the ways we are doing serialization today: (1) DDP doesn't assume that hooks are set after load, it makes sure it restores the hooks afterwards; (2) the recommend method of saving parameters using `state_dict()` doesn't save hooks, as hooks get stripped when you save `x.data` of a tensor. There are multiple places where we have implemented serialization code; all of them need to be fixed.

2. Since this is a BC-breaking change, we should warn when we previously serialized a hook, but no longer do so. After some discussion, we've decided that adding a special sentinel property to hooks will be used to suppress this warning. If a hook has the property `_torch_serialize_ignore`, we will not emit a warning if we attempt to serialize a Tensor with this hook attached to it. For example, the DDP attached hooks will have this property set.
